In [18]:
import pandas as pd


df = pd.read_csv("C:\\Users\In9.var\Desktop\Projects\data\\train (7).csv") # загружаем датасет
df

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0               842     0          2.2         0   1       0           7   
1              1021     1          0.5         1   0       1          53   
2               563     1          0.5         1   2       1          41   
3               615     1          2.5         0   0       0          10   
4              1821     1          1.2         0  13       1          44   
...             ...   ...          ...       ...  ..     ...         ...   
1995            794     1          0.5         1   0       1           2   
1996           1965     1          2.6         1   0       0          39   
1997           1911     0          0.9         1   1       1          36   
1998           1512     0          0.9         0   4       1          46   
1999            510     1          2.0         1   5       1          45   

      m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  \
0       0.6        188        2  ...         20       756  2549     9     7   
1       0.7        136        3  ...        905      1988  2631    17     3   
2       0.9        145        5  ...       1263      1716  2603    11     2   
3       0.8        131        6  ...       1216      1786  2769    16     8   
4       0.6        141        2  ...       1208      1212  1411     8     2   
...     ...        ...      ...  ...        ...       ...   ...   ...   ...   
1995    0.8        106        6  ...       1222      1890   668    13     4   
1996    0.2        187        4  ...        915      1965  2032    11    10   
1997    0.7        108        8  ...        868      1632  3057     9     1   
1998    0.1        145        5  ...        336       670   869    18    10   
1999    0.9        168        6  ...        483       754  3919    19     4   

      talk_time  three_g  touch_screen  wifi  price_range  
0            19        0             0     1            1  
1             7        1             1     0            2  
2             9        1             1     0            2  
3            11        1             0     0            2  
4            15        1             1     0            1  
...         ...      ...           ...   ...          ...  
1995         19        1             1     0            0  
1996         16        1             1     1            2  
1997          5        1             1     0            3  
1998         19        1             1     1            0  
1999          2        1             1     1            3  

[2000 rows x 21 columns]

In [19]:
df.groupby('price_range').count() # число каждого класса. прекрасные данные. спасибо!

battery_power  blue  clock_speed  dual_sim   fc  four_g  \
price_range                                                            
0                      500   500          500       500  500     500   
1                      500   500          500       500  500     500   
2                      500   500          500       500  500     500   
3                      500   500          500       500  500     500   

             int_memory  m_dep  mobile_wt  n_cores   pc  px_height  px_width  \
price_range                                                                    
0                   500    500        500      500  500        500       500   
1                   500    500        500      500  500        500       500   
2                   500    500        500      500  500        500       500   
3                   500    500        500      500  500        500       500   

             ram  sc_h  sc_w  talk_time  three_g  touch_screen  wifi  
price_range                                                           
0            500   500   500        500      500           500   500  
1            500   500   500        500      500           500   500  
2            500   500   500        500      500           500   500  
3            500   500   500        500      500           500   500

In [20]:
X = df.drop(['price_range'], axis=1)
y = df['price_range']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=42)

In [21]:
from sklearn.neighbors import KNeighborsClassifier as knn

model1 = knn()
model1.classes_ = 4
model1.fit(X_train, y_train)
model1.score(X_test, y_test) # создаю модель knn, делаю кол-во классов 4, обучаю, тестирую. Совсем неплохой результат

0.9220389805097451

In [22]:
from catboost import CatBoostClassifier as cbc, Pool

scores = []

for it in range(1, 5):
    for de in range(1, 5):
        for lr in range(1, 11):
            model2 = cbc(iterations=it, depth=de, learning_rate=lr/10, loss_function='MultiClass', verbose=True, classes_count=4)
            model2.fit(X_train, y_train)
            scores.append(model2.score(X_test, y_test))
max(scores) # Создаю catboost для классификации, перебираем циклом параметры модели, выбираем лучшее из получившихся

0:	learn: 1.3235039	total: 768us	remaining: 0us
0:	learn: 1.2673312	total: 704us	remaining: 0us
0:	learn: 1.2178461	total: 772us	remaining: 0us
0:	learn: 1.1750183	total: 736us	remaining: 0us
0:	learn: 1.1387113	total: 715us	remaining: 0us
0:	learn: 1.1086867	total: 722us	remaining: 0us
0:	learn: 1.0846161	total: 716us	remaining: 0us
0:	learn: 1.0661003	total: 723us	remaining: 0us
0:	learn: 1.0526930	total: 734us	remaining: 0us
0:	learn: 1.0439236	total: 731us	remaining: 0us
0:	learn: 1.2775220	total: 1.01ms	remaining: 0us
0:	learn: 1.1806508	total: 996us	remaining: 0us
0:	learn: 1.0962279	total: 992us	remaining: 0us
0:	learn: 1.0244970	total: 962us	remaining: 0us
0:	learn: 0.9653593	total: 1.03ms	remaining: 0us
0:	learn: 0.9183773	total: 1.24ms	remaining: 0us
0:	learn: 0.8828210	total: 946us	remaining: 0us
0:	learn: 0.8577449	total: 1.02ms	remaining: 0us
0:	learn: 0.8420765	total: 1.1ms	remaining: 0us
0:	learn: 0.8347012	total: 973us	remaining: 0us
0:	learn: 1.2800291	total: 4.67ms	re

1:	learn: 0.7328553	total: 2.16ms	remaining: 1.08ms
2:	learn: 0.6564496	total: 3.3ms	remaining: 0us
0:	learn: 0.8828210	total: 959us	remaining: 1.92ms
1:	learn: 0.6993350	total: 1.9ms	remaining: 947us
2:	learn: 0.6265206	total: 2.75ms	remaining: 0us
0:	learn: 0.8577449	total: 1.03ms	remaining: 2.07ms
1:	learn: 0.6780316	total: 2.6ms	remaining: 1.3ms
2:	learn: 0.6153919	total: 3.48ms	remaining: 0us
0:	learn: 0.8420765	total: 920us	remaining: 1.84ms
1:	learn: 0.6677903	total: 2.11ms	remaining: 1.05ms
2:	learn: 0.5913014	total: 3.07ms	remaining: 0us
0:	learn: 0.8347012	total: 931us	remaining: 1.86ms
1:	learn: 0.6773603	total: 1.85ms	remaining: 923us
2:	learn: 0.6074342	total: 2.75ms	remaining: 0us
0:	learn: 1.2800291	total: 1.22ms	remaining: 2.43ms
1:	learn: 1.1963139	total: 2.54ms	remaining: 1.27ms
2:	learn: 1.1107316	total: 3.6ms	remaining: 0us
0:	learn: 1.1850454	total: 1.32ms	remaining: 2.65ms
1:	learn: 1.0601657	total: 2.53ms	remaining: 1.27ms
2:	learn: 0.9377463	total: 3.7ms	remaini

1:	learn: 0.6742165	total: 2.62ms	remaining: 2.62ms
2:	learn: 0.5175437	total: 4.05ms	remaining: 1.35ms
3:	learn: 0.4729167	total: 5.5ms	remaining: 0us
0:	learn: 0.8343943	total: 2.61ms	remaining: 7.82ms
1:	learn: 0.6706013	total: 3.89ms	remaining: 3.89ms
2:	learn: 0.5009759	total: 5.13ms	remaining: 1.71ms
3:	learn: 0.4907069	total: 6.26ms	remaining: 0us
0:	learn: 1.2681105	total: 3.9ms	remaining: 11.7ms
1:	learn: 1.1650642	total: 5.65ms	remaining: 5.65ms
2:	learn: 1.0821531	total: 7.29ms	remaining: 2.43ms
3:	learn: 1.0173207	total: 8.95ms	remaining: 0us
0:	learn: 1.1624824	total: 5ms	remaining: 15ms
1:	learn: 1.0054569	total: 8.3ms	remaining: 8.3ms
2:	learn: 0.8909483	total: 11.4ms	remaining: 3.81ms
3:	learn: 0.8152714	total: 14.3ms	remaining: 0us
0:	learn: 1.0701476	total: 5.11ms	remaining: 15.3ms
1:	learn: 0.8823869	total: 8.62ms	remaining: 8.62ms
2:	learn: 0.7719233	total: 11.6ms	remaining: 3.87ms
3:	learn: 0.6943498	total: 14.3ms	remaining: 0us
0:	learn: 0.9915386	total: 1.74ms	re

0.8335832083958021

In [23]:
from keras.models import Sequential
from keras.layers import Dense
import numpy

numpy.random.seed(42)

In [37]:
X = df.iloc[:,0:6]
y = df.iloc[:,20]
scores=[]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=42)

for i in range(4, 15):
    for j in range(4, 10):
        for k in range(4, 10):
            for l in range(4, 10):
                model3 = Sequential([
                    Dense(i, input_dim=6, activation='sigmoid'),
                    Dense(j, activation='sigmoid'),
                    Dense(k, activation='sigmoid'),
                    Dense(l, activation='sigmoid'),
                    Dense(1, activation='softmax')
                ])
                model3.compile(loss="categorical_crossentropy", optimizer="adam", metrics=['accuracy'])
                model3.fit(X_train, y_train, epochs = 0)
                scores.append(model3.evaluate(X_test, y_test))
# подгоняем данные под 6 входных параметров, по циклам меняем кол-во нейронов в скрытых слоях, собираем сеть, считаем accuracy
max(scores) # поверьте мне accuracy: 0.2474 не меняется от кол-ва нейронов

21/21 [==============================] - 0s 617us/step - loss: 0.0000e+00 - accuracy: 0.2474


KeyboardInterrupt: 

Ответы на вопросы: 
1) датасет: 2000 rows × 21 columns
2) датасет имеет набор из 4 классов по 500 значений каждый
3) классификация - определение принадлежности к тому или иному классу(да грубовато)
4) ансамбль - использование взаимодействующих друг с другом моделей, чаще всего базовых
5) нейронная сеть - это искусственное воссоздание работы мозга для решения определенных задач
6) метрики классификации - accuracy, precision, recall, f1-score
Наилучшим показал себя KNN - 0.9220389805097451. Ансамбль не многим хуже - 0.8335832083958021. В принципе не нужно было решать эту задачу сетью. Как и ожидалось она показала себя в разы хуже остальных моделей - 0.2474.